storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1212 @@
+		return -EIO;
+	}
+
+	if (CARD_STATUS_CURRENT_STATE(finalStatus) != CARD_STATUS_CURRENT_STATE_TRAN) {
+		/* Something unexpected must have happened because card is not in a state to transfer data */
+		return -EIO;
+	}
+
+	return fallbackMode ? 0 : sdcard_wideAndFast(host);
+}
+
+
+/* Extract information about supported functions in a given function group from the function register */
+static inline uint16_t sdcard_extractFunctionGroupInfo(uint8_t fnRegister[64], uint8_t fnGroup)
+{
+	uint8_t fnGroupIndex;
+	if ((fnGroup > 6) || (fnGroup < 1)) {
+		return 0;
+	}
+
+	fnGroupIndex = (6 - fnGroup) * 2 + 2;
+	return ((uint16_t)fnRegister[fnGroupIndex] << 8) | fnRegister[fnGroupIndex + 1];
+}
+
+
+/* Extract status of function from a given function group from the function register */
+static inline uint8_t sdcard_extractFunctionSwitchResult(uint8_t fnRegister[64], uint8_t fnGroup)
+{
+	uint8_t fnGroupIndex, fnGroupBits;
+	if ((fnGroup > 6) || (fnGroup < 1)) {
+		return 0;
+	}
+
+	fnGroup = (6 - fnGroup);
+	fnGroupIndex = fnGroup / 2 + 14;
+	fnGroupBits = (fnGroup % 2 == 0) ? 4 : 0;
+	return (fnRegister[fnGroupIndex] >> fnGroupBits) & 0xF;
+}
+
+
+/* Switches card to maximum width and speed supported */
+static int sdcard_wideAndFast(sdcard_hostData_t *host)
+{
+	uint8_t bigRegs[64];
+	uint32_t switchFunctionArg;
+	uint16_t accessModeFunctions;
+
+	/* Get SD card configuration register with some useful info about the card */
+	if (sdio_cmdSendEx(host, SDIO_ACMD51_SEND_SCR, 0, NULL, 0, bigRegs) < 0) {
+		return -EIO;
+	}
+
+	bool cmd6Supported = SCR_SD_SPEC(bigRegs) >= SCR_SD_SPEC_V1_10;
+	/* In theory all SD cards should support 4-bit, but make sure */
+	if ((SCR_BUS_WIDTHS(bigRegs) & SCR_BUS_WIDTHS_4_BIT) != 0) {
+		if (sdio_cmdSend(host, SDIO_ACMD6_SET_BUS_WIDTH, 2, NULL) < 0) {
+			LOG_ERROR("bus widening failed");
+			return -EIO;
+		}
+
+		*(host->base + SDHOST_REG_HOST_CONTROL) |= HOST_CONTROL_4_BIT_MODE;
+		usleep(10);
+		if (sdio_cmdSendEx(host, SDIO_ACMD13_SD_STATUS, 0, NULL, 0, bigRegs) < 0) {
+			LOG_ERROR("bus widening failed");
+			return -EIO;
+		}
+
+		if (SD_STATUS_DAT_BUS_WIDTH(bigRegs) != SD_STATUS_DAT_BUS_WIDTH_4_BIT) {
+			LOG_ERROR("bus widening failed");
+			return -EIO;
+		}
+	}
+
+	bool isHighSpeedSupported = false;
+	if (cmd6Supported) {
+		switchFunctionArg = SDIO_SWITCH_FUNC_GET | SDIO_SWITCH_FUNC_HIGH_SPEED;
+		if (sdio_cmdSendEx(host, SDIO_CMD6_SWITCH_FUNC, switchFunctionArg, NULL, 0, bigRegs) == 0) {
+			accessModeFunctions = sdcard_extractFunctionGroupInfo(bigRegs, SDCARD_FUNCTION_GROUP_ACCESS_MODE);
+			if ((accessModeFunctions & SDCARD_FUNCTION_GROUP_ACCESS_MODE_HIGH_SPEED) != 0) {
+				switchFunctionArg |= SDIO_SWITCH_FUNC_SET;
+				if (sdio_cmdSendEx(host, SDIO_CMD6_SWITCH_FUNC, switchFunctionArg, NULL, 0, bigRegs) == 0) {
+					if (sdcard_extractFunctionSwitchResult(bigRegs, SDCARD_FUNCTION_GROUP_ACCESS_MODE) == 1) {
+						isHighSpeedSupported = true;
+					}
+				}
+			}
+		}
+	}
```

Comment:
I can't think of MISRA-friendly way to combat this indentation level. As we ignore early return rule anyway, perhaps a function?